mllib-local/src/main/scala/org/apache/spark/ml/linalg/Vectors.scala

```diff
@@ -370,14 +370,19 @@ object Vectors {
       case (v1: DenseVector, v2: SparseVector) =>
         squaredDistance = sqdist(v2, v1)
 
-      case (DenseVector(vv1), DenseVector(vv2)) =>
-        var kv = 0
+      case (DenseVector(vv1), DenseVector(vv2)) => {
         val sz = vv1.length
-        while (kv < sz) {
-          val score = vv1(kv) - vv2(kv)
-          squaredDistance += score * score
-          kv += 1
+        @annotation.tailrec
```

Comment:
Why would tail recursion be faster? I can't really imagine why. A small method can get JITted, but the overhead of the method calls is still got to be much higher than an increment and compare. I'd like to understand how it was benchmarked and what the difference is

Reply:
FWIW, I did an experimental by myself before because someone proposed similar change (turning `while` to tail recursive one). There was no virtual difference. If this change is only tail recursion vs `while`, I doubt how much it can improve. I believe the benchmark steps should be clarified.

Reply:
Hi, i put it [here](https://github.com/beckgael/functional_sqdist_spark) my experimental process with a SparkNotebook or with sbt project in order to allow others to tests.
I took the `sqdist` function as it is, make a functional version of it and compare both as best as i can.
Run the notebook on 100k pair (Imperativ/Functional) of `sqdist` for medium size vectors took less than 1min and results are different enough to consider them i presume, even if it's not the best way to test it and i'm sorry for that.
After having browsed on StackOverFlow i fall on [this one](https://stackoverflow.com/questions/9168624/why-is-my-scala-tail-recursion-faster-than-the-while-loop) which talks about a difference between JVM version on tailrec.

I hope we'll find something interesting and that i'm not disturbing you on your fantastic work !

Reply:
The StackOverflow post points to a reason the two implementations that were being compared there could be different. Here I don't see any branching is saved. 

I can't explain your microbenchmark, which still surprises me. If I had to guess it's because of JIT compilation or something. I do wonder whether the microbenchmark shows an improvement that won't actually materialize in this code. That is, I think you'd have to show that this improves the performance of Spark DenseVectors when used as-is in the Vector class here.

I'm pretty skeptical but stranger things have happened.